Support both iso639-3 codes and BCP-47 language tags

ekaf opened the pull request:
Fix #3058 by adding a function to obtain language names, using their respective iso639-3 language codes.

This PR adds a langname.py module in the nltk directory. I plan to use it with wordnet.py, to obtain the names of the multilingual wordnets from OMW.

```
>>>from nltk.langnames import langname
>>>for code in ["cat", "wln"]:
>>>    print(f"{code}: {langname(code)}")
```

> cat: Catalan
> wln: Walloon
> 

The language codes were obtained from https://iso639-3.sil.org/code_tables/download_tables
and since it is not clear whether this use is authorized,  it seems safer to ask them:

> For any questions about whether a particular use is covered by these guidelines, contact the Registration Authority at ISO639-3@sil.org.

ekaf committed:
Add support for ISO-639-3 language codes

ekaf committed:
Add langname() function with doctest

ekaf committed:
Add file header

ekaf committed:
Revert "Add langname() function with doctest"

This reverts commit f2b9ce4171bf028d690cfc0b1126d8b3d32d0947.

ekaf committed:
Revert "Add file header"

This reverts commit cbfc2ec5c0c59f005c133b9ad141c5343bbc26a6.

ekaf committed:
Revert "Add support for ISO-639-3 language codes"

This reverts commit 46fb31b5ca5277333a99846141eb0df8095b3f61.

ekaf committed:
Merge remote-tracking branch 'upstream/develop' into develop

ekaf committed:
Add support for iso639-3 language codes

ekaf commented:
@tomaarsen, curiously CI only fails with Python 3.9 on Ubuntu, but succeeds on the other platforms. The failure seems unrelated with this PR, since all the errors concern _nltk/parse/corenlp.py_:

`grep -i err 'Python 3.9 on ubuntu-latest/9_Run pytest.txt'`

```
2022-10-05T11:48:57.8556549Z ==================================== ERRORS ====================================
2022-10-05T11:48:57.8557014Z _______________ ERROR at setup of TestTokenizerAPI.test_tokenize _______________
2022-10-05T11:48:57.8597206Z E       OSError: [Errno 98] Address already in use
2022-10-05T11:48:57.8599517Z nltk/parse/corenlp.py:32: OSError
2022-10-05T11:48:57.8601953Z _______________ ERROR at setup of TestTaggerAPI.test_ner_tagger ________________
2022-10-05T11:48:57.8635756Z E       OSError: [Errno 98] Address already in use
2022-10-05T11:48:57.8636117Z nltk/parse/corenlp.py:32: OSError
2022-10-05T11:48:57.8657887Z _______________ ERROR at setup of TestTaggerAPI.test_pos_tagger ________________
2022-10-05T11:48:57.8739007Z E       OSError: [Errno 98] Address already in use
2022-10-05T11:48:57.8741508Z nltk/parse/corenlp.py:32: OSError
2022-10-05T11:48:57.8744008Z ___________ ERROR at setup of TestTaggerAPI.test_unexpected_tagtype ____________
2022-10-05T11:48:57.8766761Z E       OSError: [Errno 98] Address already in use
2022-10-05T11:48:57.8769194Z nltk/parse/corenlp.py:32: OSError
2022-10-05T11:48:57.8769611Z ____________ ERROR at setup of TestParserAPI.test_dependency_parser ____________
2022-10-05T11:48:57.8823049Z E       OSError: [Errno 98] Address already in use
2022-10-05T11:48:57.8881010Z nltk/parse/corenlp.py:32: OSError
2022-10-05T11:48:57.8886635Z __________________ ERROR at setup of TestParserAPI.test_parse __________________
2022-10-05T11:48:57.8976601Z E       OSError: [Errno 98] Address already in use
2022-10-05T11:48:57.8981665Z nltk/parse/corenlp.py:32: OSError
2022-10-05T11:48:57.9228352Z = 697 passed, 26 skipped, 9 xfailed, 18 warnings, 6 errors in 127.83s (0:02:07) =
2022-10-05T11:48:58.0065390Z ##[error]Process completed with exit code 1.



```

tomaarsen commented:
I'm noticing the test failures. Out of Windows, Mac and Ubuntu, only the Ubuntu test setup prepares the download of third-party tools, including CoreNLP. The Windows and Mac suites simply skip these tests if it is discovered that the relevant jar files are not present.

The issue seems to indicate that either:
* The `CoreNLPServer` instance does not always correctly terminate, causing future tests to fail when they try to claim port 9000.
* The CI runner attempts to run tests in parellel in such a way that two separate instances of `CoreNLPServer` are attempted to be started on 9000 at the same time.

I'll be looking into this issue, and otherwise I'll simply set these 6 tests to always be skipped. The test failures can be ignored in this PR.

ekaf commented:
@tomaarsen, ok but it still seems mysterious that Ubuntu only fails with Python 3.9, and succeeds with the other Python versions.

tomaarsen commented:
They fail arbitrarily. Initially, they failed for all 4 versions. Then, I restarted the failing ones, and 3 of the 4 passed, leaving only the Python 3.9 one to fail.

ekaf committed:
Add support for retired language codes

ekaf commented:
Some language codes used by the Extended OMW were retired in the latest version of the iso639-3 standard. Now, this PR also supports the retired codes, so we can get a language name for all the wordnets in OMW-1.4 plus the Extended OMW (1222 wordnets, slightly fewer languages):

```

from nltk.corpus import wordnet as wn
from nltk.langnames import langname

wn.add_omw()
wn.add_exomw()

for lang in wn.langs():
    code = lang.split("_")[0]
    print(f"{lang}: {langname(code)}")

```

> 
> eng: English
> als: Tosk Albanian
> arb: Standard Arabic
> bul: Bulgarian
> cmn: Mandarin Chinese
> [...]
> zro_wikt: Záparo
> zsm_wikt: Standard Malay
> zul_wikt: Zulu
> zun_wikt: Zuni
> zza_wikt: Zaza
>

stevenbird commented:
I think generic functionality like language identity belongs in a top-level module.

ekaf committed:
Move langnames.py to the top-level

ekaf commented:
The hardest part may be to obtain approval from _iso639-3.sil.org_.
I can ask them, unless somebody else wants to.

ekaf commented:
An alternative could be to ask users to download the data from iso639-3.sil.org, but this seems more cumbersome, since it would require checking at each call that the functionality is available. Any suggestions?

tomaarsen commented:
This functionality seems like a convenience function, that someone should be able to quickly import and run a language identifier through. If the user would have to first download some file from iso639-3.sil.org, then it'll be considerably easier for them to just use their search engine of choice to quickly find out what the language code means. 
So I would rather not go that route.

ekaf committed:
Add langcode() function

ekaf commented:
This use of the iso639-3 data seems to conform to the registration authority's guidelines. Here's their response:

> From: Registrar ISO639-3 <iso639-3@sil.org>
> Date: Fri, 7 Oct 2022 11:50:45 -0500
> Subject: Re: Terms of use
> 
> Dear Eric,
> 
> Your use described in this proposal appears to conform to our terms of use
> <https://iso639-3.sil.org/code_tables/download_tables#termsofuse>.  You may
> request to be informed
> <https://iso639-3.sil.org/code_changes/requesting_notification_of_changes>
> about updates to the download tables, as we do not have an API, and there
> are updates to the tables posted annually.  In addition, we anticipate that
> changes be approved on a more frequent basis starting in the next year.
> 
> Kind regards,
> 
> Janell
>

ekaf committed:
Add iso639retired dictionary

ekaf commented:
Parenthesized text in the language names should not be discarded, because it would make the _iso639name_ dictionary surjective, thus  breaking the bijectivity of the mapping. This also happens if the retired codes are added to the same dictionary. 

The problem is fixed now, by keeping the retired codes in a separate _iso639retired_  dictionary, and also keeping parenthesized text.

ekaf committed:
Improve wrapper functions

ekaf commented:
Although each dictionary here is a bijective mapping, the wrapper functions, which combine a main dictionary with one for retired codes, deviate from bijectivity in the cases where the code for a particular language is retired and replaced by a new code for the same language. 
In these cases, the _langcode()_ wrapper function return only the new iso639-3 code, not the retired code: 

```
import nltk.langnames as lgn

for (ret_code, ret_name) in lgn.iso639retired.items():
    code = lgn.langcode(ret_name)
    if ret_code != code:
        print(f"{ret_code}:{code}:{ret_name}")

```

> 
> fri:fry:Western Frisian
> cit:ctg:Chittagonian
> flm:cfm:Falam Chin
> nhs:npl:Southeastern Puebla Nahuatl
> aiz:aiw:Aari
> azr:adz:Adzera
> bcx:pmf:Pamona
> bii:bzi:Bisu
> blu:hnj:Hmong Njua
> fiz:izr:Izere
> kds:lhi:Lahu Shi
> mdo:gso:Southwest Gbaya
> nky:kix:Khiamniungan Naga
> ork:okv:Orokaiva
> rjb:rjs:Rajbanshi
> suf:tpf:Tarpia
> suh:sxb:Suba
> mly:zlm:Malay (individual language)
> muw:unr:Mundari
> xst:stv:Silt'e
> scc:srp:Serbian
> scr:hrv:Croatian
> agp:prf:Paranan
> rmr:rmq:Caló
> sul:sgd:Surigaonon
> wgw:wgb:Wagawaga
> bjq:bzc:Southern Betsimisaraka Malagasy
> nbf:nxq:Naxi
> baz:tvu:Tunen
> kpp:jkp:Paku Karen
> wiw:wgu:Wirangu
> yen:ynq:Yendang
> daf:dnj:Dan
> djl:dze:Djiwarli
> nlr:nrk:Ngarla
> wit:wnw:Wintu
> yiy:yyr:Yir Yoront
> sap:spn:Sanapaná
> duj:dwu:Dhuwal
> kjf:klj:Khalaj
> kxu:uki:Kui (India)
> sdm:sdq:Semandang
> gji:gyz:Geji
> lno:lgo:Lango (South Sudan)
> wya:wyn:Wyandot
>

ekaf committed:
Add module docstring with doctest

ekaf committed:
Add 2-letter language codes

ekaf commented:
@tomaarsen, CI failed with the following error in _pre-commit_, that did not occur with the previous Python version (3.10.7). Could it indicate a problem with Python 3.10.8? Then, maybe nothing could pass pre-commit with this setup. 
```
2022-10-27T11:47:13.1712090Z Traceback (most recent call last):
2022-10-27T11:47:13.1712998Z   File "/home/runner/.cache/pre-commit/repoaro_9qfm/py_env-python3.10/bin/pyupgrade", line 5, in <module>
2022-10-27T11:47:13.1713718Z     from pyupgrade._main import main
2022-10-27T11:47:13.1714430Z   File "/home/runner/.cache/pre-commit/repoaro_9qfm/py_env-python3.10/lib/python3.10/site-packages/pyupgrade/_main.py", line 30, in <module>
2022-10-27T11:47:13.1715170Z     from pyupgrade._data import FUNCS
2022-10-27T11:47:13.1715867Z   File "/home/runner/.cache/pre-commit/repoaro_9qfm/py_env-python3.10/lib/python3.10/site-packages/pyupgrade/_data.py", line 126, in <module>
2022-10-27T11:47:13.1716585Z     _import_plugins()
2022-10-27T11:47:13.1717310Z   File "/home/runner/.cache/pre-commit/repoaro_9qfm/py_env-python3.10/lib/python3.10/site-packages/pyupgrade/_data.py", line 123, in _import_plugins
2022-10-27T11:47:13.1718134Z     __import__(name, fromlist=['_trash'])
2022-10-27T11:47:13.1718862Z   File "/home/runner/.cache/pre-commit/repoaro_9qfm/py_env-python3.10/lib/python3.10/site-packages/pyupgrade/_plugins/pep584.py", line 5, in <module>
2022-10-27T11:47:13.1719614Z     from tokenize_rt import List
2022-10-27T11:47:13.1720341Z ImportError: cannot import name 'List' from 'tokenize_rt' (/home/runner/.cache/pre-commit/repoaro_9qfm/py_env-python3.10/lib/python3.10/site-packages/tokenize_rt.py)
2022-10-27T11:47:13.1721014Z 
2022-10-27T11:47:33.5007538Z black....................................................................[42mPassed[m
2022-10-27T11:47:34.9601338Z isort....................................................................[42mPassed[m
2022-10-27T11:47:34.9776285Z ##[error]The process '/opt/hostedtoolcache/Python/3.10.8/x64/bin/pre-commit' failed with exit code 1

```

ekaf committed:
Add regular expression check

ekaf commented:
@tomaarsen, I noticed that I had  committed some unwanted files, so I reset this branch, which may accidentally have overwritten your upgrade of pyupgrade (I am not too sure of this). Anyway, the error seems the same as before:
```
2022-10-28T10:38:43.9604962Z Traceback (most recent call last):
2022-10-28T10:38:43.9605731Z   File "/home/runner/.cache/pre-commit/repo825n37x5/py_env-python3.10/bin/pyupgrade", line 5, in <module>
2022-10-28T10:38:43.9606077Z     from pyupgrade._main import main
2022-10-28T10:38:43.9606652Z   File "/home/runner/.cache/pre-commit/repo825n37x5/py_env-python3.10/lib/python3.10/site-packages/pyupgrade/_main.py", line 30, in <module>
2022-10-28T10:38:43.9607042Z     from pyupgrade._data import FUNCS
2022-10-28T10:38:43.9607610Z   File "/home/runner/.cache/pre-commit/repo825n37x5/py_env-python3.10/lib/python3.10/site-packages/pyupgrade/_data.py", line 126, in <module>
2022-10-28T10:38:43.9607974Z     _import_plugins()
2022-10-28T10:38:43.9608557Z   File "/home/runner/.cache/pre-commit/repo825n37x5/py_env-python3.10/lib/python3.10/site-packages/pyupgrade/_data.py", line 123, in _import_plugins
2022-10-28T10:38:43.9609015Z     __import__(name, fromlist=['_trash'])
2022-10-28T10:38:43.9609601Z   File "/home/runner/.cache/pre-commit/repo825n37x5/py_env-python3.10/lib/python3.10/site-packages/pyupgrade/_plugins/pep584.py", line 5, in <module>
2022-10-28T10:38:43.9610006Z     from tokenize_rt import List
2022-10-28T10:38:43.9610587Z ImportError: cannot import name 'List' from 'tokenize_rt' (/home/runner/.cache/pre-commit/repo825n37x5/py_env-python3.10/lib/python3.10/site-packages/tokenize_rt.py)
2022-10-28T10:38:43.9610895Z 
2022-10-28T10:39:06.1911460Z black....................................................................[42mPassed[m
2022-10-28T10:39:07.6552814Z isort....................................................................[42mPassed[m
2022-10-28T10:39:07.6773891Z ##[error]The process '/opt/hostedtoolcache/Python/3.10.8/x64/bin/pre-commit' failed with exit
``` code 1

tomaarsen committed:
Update pyupgrade version to avoid ImportError for Python 3.10.8

tomaarsen commented:
It has indeed. All good though.

ekaf commented:
Thanks @tomaarsen, this solved the _pre-commit_ problem. So now, we are back at the _corenlp_ issue, except that it now fails on all Ubuntu installs, where it failed more arbitrarily before.
```

nltk/test/unit/test_corenlp.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
nltk/parse/corenlp.py:80: in __init__
    try_port(port)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

port = 9000

    def try_port(port=0):
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
>       sock.bind(("", port))
E       OSError: [Errno 98] Address already in use

nltk/parse/corenlp.py:32: OSError
____________ ERROR at setup of TestParserAPI.test_dependency_parser ____________
[gw0] linux -- Python 3.7.15 /opt/hostedtoolcache/Python/3.7.15/x64/bin/python
```

tomaarsen committed:
Always skip CoreNLP doctests

ekaf committed:
Improve inverse lookup of retired codes

ekaf committed:
Merge branch 'langnames' of https://github.com/ekaf/nltk into langnames

ekaf commented:
In order to support BCP-47, _langcode_ now returns the short 2-letter code by default (pass typ=3 to get the 3-letter code).

```
import nltk.langnames as lgn

for (ret_code, ret_name) in lgn.iso639retired.items():
    code = lgn.langcode(ret_name)
    if len(ret_code) > len(code):
        print(f"{ret_code}:{code}:{ret_name}")

```

> fri:fy:Western Frisian
> occ:ie:Occidental
> scc:sr:Serbian
> scr:hr:Croatian
> mol:ro:Moldavian

ekaf committed:
Merge remote-tracking branch 'upstream/develop' into langnames

ekaf commented:
I consider removing the data dictionaries from _langnames.py_, and replacing them by a CorpusReader for the BCP-47 database (http://www.iana.org/assignments/language-subtag-registry). This would allow a much preferable separation of program and data.

Unlike the iso639-3 data, the BCP-47 database has a permissive license, and contains all the needed language codes and names (except the 3-letter codes for around 200 languages which have a 2-letter code).

Then, the generic language id functions could remain in a much shorter top-level _langnames.py_ module.

ekaf committed:
Support BCP-47

ekaf commented:
Added a _bcp47_ CorpusReader, which needs the new bcp47
nltk_data package  (https://github.com/nltk/nltk_data/pull/191).

Iso 639-3 is now supported in _langnames_py_ through bcp47, so all doctests fail if the bcp47 data package is missing.

ekaf committed:
Avoid deprecated langcodes

ekaf committed:
Merge branch 'nltk:develop' into langnames

ekaf committed:
Merge remote-tracking branch 'upstream/develop' into langnames

ekaf committed:
Fix all_synsets() function

tomaarsen committed:
Set stack level for warnings to warn on the langname call

Now it throws e.g.
```
...\nltk_3060.py:9: UserWarning: Shortening 'smo' to 'sm'
  print(f"{lang}: {langname(code)}")
```

Rather than
```
...\nltk\langnames.py:64: UserWarning: Shortening zha to za
  warn(f"Shortening {code} to {code2}")
```

tomaarsen committed:
Dict key membership is equivalent to dict membership

tomaarsen committed:
Dict key membership is equivalent to dict membership

tomaarsen committed:
Resolve bug: subtag -> tag

tomaarsen committed:
Use !r for repr formatting in warnings

tomaarsen commented on nltk/corpus/reader/bcp47.py:
```diff
@@ -0,0 +1,216 @@
+from nltk.corpus.reader import CorpusReader
+
+
+class Bcp47CorpusReader(CorpusReader):
```
```suggestion
class BCP47CorpusReader(CorpusReader):
```
It seems that BCP47 is more fitting.

ekaf replied:
I agree, but since all other corpus reader classes include CorpusReader in their name (see the \_\_all\_\_ list in nltk/corpus/reader/\_\_init\_\_.py), I prefer to do the same for the moment.

tomaarsen replied:
Apologies, I meant `BCP47CorpusReader`, i.e. capitalising Bcp to BCP.

ekaf replied:
Oh, ok, we can do that, but it will require changes in 3 files.

tomaarsen replied:
If you'd like, I can handle that, and then merge it afterwards. If you think we're ready for merging afterwards 😄

ekaf replied:
Thanks @tomaarsen, I just made the change.

tomaarsen commented on nltk/corpus/reader/bcp47.py:
```diff
@@ -0,0 +1,216 @@
+        """Convert a BCP-47 tag to a dictionary of labelled subtags"""
+        subtags = tag.split("-")
+        lang = {}
+        labels = ["language", "extlang", "script", "region", "variant", "variant"]
```
```suggestion
        labels = ["language", "extlang", "script", "region", "variant"]
```
Was the double `"variant"` on purpose?

ekaf replied:
It is indeed intended, since some language tags may specify a second "variant" field. This was however not properly handled, but it is fixed now:

```
import nltk.langnames as lgn
print(lgn.langname("oc-provenc-grmistr"))
```

> Occitan (post 1500): Provençal: Mistralian or Mistralian-inspired Occitan orthography

tomaarsen reviewed (approved):
@ekaf 

I've made some minor changes and one fix. The tests pass for me (locally), so I'm nearing the point where I'm ready to merge, although I do have some changes that still need doing at https://github.com/nltk/nltk_data/pull/191.

That said, I do have some questions, see my comments. This is looking good, though! Glad we want the NLTK data route.

tomaarsen committed:
Add simple regression tests for #3077

tomaarsen committed:
Merge branch 'develop' of https://github.com/nltk/nltk into ekaf-all_synsets

ekaf committed:
Merge commit 'refs/pull/3060/head' of https://github.com/nltk/nltk into langnames

ekaf committed:
Merge commit 'refs/pull/3078/head' of https://github.com/nltk/nltk into langnames

ekaf committed:
Eventually store a second 'variant' field

ekaf commented:
Thanks for all your changes @tomaarsen, and in particular for catching the nasty confusion between 'tag' and 'subtag'.

ekaf committed:
Capitalize BCP47 in CorpusReader name

ekaf committed:
Fix conflict with updated wordnet.py

tomaarsen committed:
Reimplement removed type hint changes from #3081

tomaarsen commented:
Wonderful! Thank you for this!
